src/trajectory.jl

```diff
@@ -110,54 +110,94 @@ Ref: https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/hmc/nuts/base_n
 
 Create a slice sampler for a single leaf tree:
 - the slice variable is copied from the passed-in sampler `s` and
 - the number of acceptable candicates is computed by comparing the slice variable against the current energy.
 """
-function SliceTS(s::SliceTS, H0::AbstractFloat, zcand::PhasePoint)
-    return SliceTS(zcand, s.ℓu, (s.ℓu <= -energy(zcand)) ? 1 : 0)
+function SliceTS(s::SliceTS, H0::AbstractScalarOrVec{<:AbstractFloat}, zcand::PhasePoint)
+    return SliceTS(zcand, s.ℓu, ifelse.(s.ℓu .<= .-energy(zcand), 1, 0))
 end
 
 """
-    MultinomialTS(s::MultinomialTS, H0::AbstractFloat, zcand::PhasePoint)
+    MultinomialTS(s::MultinomialTS, H0::AbstractScalarOrVec{<:AbstractFloat}, zcand::PhasePoint)
 
 Multinomial sampler for a trajectory consisting only a leaf node.
 - tree weight is the (unnormalised) energy of the leaf.
 """
-function MultinomialTS(s::MultinomialTS, H0::AbstractFloat, zcand::PhasePoint)
-    return MultinomialTS(zcand, H0 - energy(zcand))
+function MultinomialTS(s::MultinomialTS, H0::AbstractScalarOrVec{<:AbstractFloat}, zcand::PhasePoint)
+    return MultinomialTS(zcand, H0 .- energy(zcand))
 end
 
-function combine(rng::AbstractRNG, s1::SliceTS, s2::SliceTS)
+function combine(rng::AbstractRNG, s1::SliceTS{FT,FT}, s2::SliceTS{FT,FT}) where {FT<:AbstractFloat}
     @assert s1.ℓu == s2.ℓu "Cannot combine two slice sampler with different slice variable"
     n = s1.n + s2.n
     zcand = rand(rng) < s1.n / n ? s1.zcand : s2.zcand
     return SliceTS(zcand, s1.ℓu, n)
 end
+function combine(rng::AbstractRNG, s1::SliceTS, s2::SliceTS)
+    @assert s1.ℓu == s2.ℓu "Cannot combine two slice sampler with different slice variable"
+    n = s1.n + s2.n
+    # ensure is_accept size matches s1.ℓu
+    α = s1.n / n
+    is_accept = broadcast(s1.ℓu) do _
+        rand(rng) < α
+    end
+    zcand = deepcopy(s1.zcand)
+    zcand = accept!(s2.zcand, zcand, is_accept)
+    return SliceTS(zcand, s1.ℓu, n)
+end
 
 function combine(zcand::PhasePoint, s1::SliceTS, s2::SliceTS)
     @assert s1.ℓu == s2.ℓu "Cannot combine two slice sampler with different slice variable"
     n = s1.n + s2.n
     return SliceTS(zcand, s1.ℓu, n)
 end
 
-function combine(rng::AbstractRNG, s1::MultinomialTS, s2::MultinomialTS)
+function accept!(
+    s::SliceTS,
+    s′::SliceTS,
+    is_accept::AbstractVector{Bool}
+)
+    zcand = accept!(s.zcand, s′.zcand, is_accept)
+    ℓu = accept!(s.ℓu, s′.ℓu, is_accept)
+    n = accept!(s.n, s′.n, is_accept)
+    return SliceTS(zcand, ℓu, n)
+end
+
+function combine(rng::AbstractRNG, s1::MultinomialTS{FT,FT}, s2::MultinomialTS{FT,FT}) where {FT<:AbstractFloat}
     ℓw = logaddexp(s1.ℓw, s2.ℓw)
     zcand = rand(rng) < exp(s1.ℓw - ℓw) ? s1.zcand : s2.zcand
     return MultinomialTS(zcand, ℓw)
 end
+function combine(rng::AbstractRNG, s1::MultinomialTS, s2::MultinomialTS)
+    ℓw = logaddexp.(s1.ℓw, s2.ℓw)
+    is_accept = rand.(rng) .< exp.(s1.ℓw .- ℓw)
```

Comment:
Why do we have the broadcast `rand` for `rng` here even we have `rng:: AbstractRNG`?